### PR TITLE
Add debug option in settings to export notes to JSON file

### PR DESCRIPTION
**Pull request by dmsnell**
This is a build to debug a specific issue in a support ticket

Creates a new item in the settings to export the account to a JSON object which can be used for further diagnosis or analysis.

Screen Shot 2020-05-02 at 9 49 17 PM

This PR is meant to explore the idea and provide assistance for devices in specific circumstances and so it's not primarily intended to merge into the application. Therefore the code is hacked together and there is no concept of separating data concerns.

**Testing**

Login to Simplenote
Open Settings
Click on `Export account` at end of view
Save file to device
Inspect file

### CODE REVIEW

**Comment by mokagio** [comment]
Bumping the milestone for this PR since it's clearly not a requirement for 2.7, which I'm about to code freeze.

>DO NOT MERGE - this is a build to debug a specific issue in a support ticket

**Comment by theck13** [comment]
I updated the file format to match the export from the web app, moved **_Export data_** to the **_Account_** section in **_Settings_**, and made some minor updates to help these changes be included in a release build.  The file can be saved to the local device storage or another storage service like Google Drive.  See the animation below for illustration.

<kbd><a href="https://user-images.githubusercontent.com/3827611/85188677-bff27380-b265-11ea-8961-a973dd45432c.gif"></a></kbd>

@dmsnell, would you be able to review my changes?  @SylvesterWilmott, are the location of **_Export data_** and the success/failure messages (i.e. "Data saved to file" and "Data could not be saved") alright?

**Comment by SylvesterWilmott** [comment]
>Are the location of Export data (i.e. under the Account section) and the success/failure messages (i.e. "Data saved to file" and "Data could not be saved") alright?

LGTM!

**Comment by mokagio** [comment]
Bumping the milestone for this PR since it's clearly not a requirement for 2.8, which I'm about to code freeze.

> DO NOT MERGE - this is a build to debug a specific issue in a support ticket

I noticed that this looks like it's getting close to be ready to merge. If you want to ship it with 2.8, I'll push a new beta once this is ready. In that case please change the base to be `release/2.8`.

File: Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java [hunk 1]
```diff
@@ -197,6 +216,83 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+            JSONObject account = new JSONObject();
+            Bucket.ObjectCursor<Note> cursor = noteBucket.allObjects();
+
+            try {
+                JSONArray activeNotes = new JSONArray();
+                JSONArray trashedNotes = new JSONArray();
+                Comparator<String> comparator = new Comparator<String>() {
+                    @Override
+                    public int compare(String text1, String text2) {
+                        return text1.compareToIgnoreCase(text2);
```
Comment (dmsnell) [inline Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java:237]: for whatever value it has we're not sorting by a locale-sensitive sort here. given the complexity of doing locale-aware sorts _and_ locale-aware lower-casing and given that it only controls the tag list in the output I'm sure it's fine. at that though, do we need them sorted?

File: Simplenote/src/main/java/com/automattic/simplenote/models/Note.java [hunk 2]
```diff
@@ -144,6 +148,15 @@ public static Calendar numberToDate(Number time) {
         return date;
     }
 
+    public static String numberToDateString(@NonNull Number number) {
+        long milliseconds = new BigDecimal(number.toString()).multiply(new BigDecimal(1000)).longValue();
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTimeInMillis(milliseconds);
+        SimpleDateFormat date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US);
+        date.setTimeZone(TimeZone.getTimeZone("UTC"));
+        return date.format(calendar.getTime());
```
Comment (dmsnell) [inline Simplenote/src/main/java/com/automattic/simplenote/models/Note.java:157]: looks like we may be able to use `Instant` if it's available

```java
Instant.ofEpochSecond(number).toString() // formatted in ISO8601 format
```

File: Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java (hunk 1 above)
Comment (theck13) [reply Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java:237]: They don't _need_ to be sorted.  I just adding sorting to mimic the web app export file.

File: Simplenote/src/main/java/com/automattic/simplenote/models/Note.java (hunk 2 above)
Comment (theck13) [reply Simplenote/src/main/java/com/automattic/simplenote/models/Note.java:157]: Nice find!  [`Instant`](https://developer.android.com/reference/java/time/Instant) requires API 26 and our minimum SDK version is 23.  Once we updated `minSdkVersion` to 26, we should definitely use `Instant.ofEpochSecond(number).toString()`.

**Comment by theck13** [comment]
Thanks for doing the heavy lifting on this feature, @dmsnell!

The one pending question is about sorting and whether we should do it or not.  Let me know if you approve these changes and we can merge this pull request.

**Comment by dmsnell** [comment]
This looks great to me.

Concerning sorting I don't think it matters so much if we do it or not. I do think though that if we choose to sort it then it would be best to have some stable cross-platform ordering but really, it doesn't matter IMO.

✅ 🚢 :shipit:

**Pull request merged**